### PR TITLE
Fix operators tokenize order

### PR DESCRIPTION
**Pull request by adrianlee44**
Also added tests for tokenizing operators

The problem with the original grammar was that it was tokenizing compound operators, such as `+=`, `-=`, `!=` after single operators.

This reason in `!=` being tokenize as `!` and `=`.

I also made the regex more compact.

I'm hoping I got all the operators in the tests but having another pair of eyes going through the changes would be great.

### CODE REVIEW

File: grammars/coffeescript.cson [hunk 1]
```diff
@@ -143,7 +143,7 @@
     'name': 'keyword.control.coffee'
   }
   {
-    'match': 'and=|or=|!|%|&|\\^|\\*|\\/|(\\-)?\\-(?!>)|\\+\\+|\\+|~|==|=(?!>)|!=|<=|>=|<<=|>>=|>>>=|<>|<|>|!|&&|\\.\\.(\\.)?|\\?|\\||\\|\\||\\:|\\*=|(?<!\\()/=|%=|\\+=|\\-=|&=|\\^=|\\b(?<![\\.\\$])(instanceof|new|delete|typeof|and|or|is|isnt|not|super)(?!\\s*:)\\b'
+    'match': '(?:(?:and|or|<<|>>>?|(?<!\\()\\/|[=!<>*%+\\-&^])?=(?!>)|[!%^*\\/~?:]|\\-?\\-(?!>)|\\+\\+?|<>|<|>|&&?|\\.\\.\\.?|\\|\\|?|\\b(?<![\\.\\$])(instanceof|new|delete|typeof|and|or|is|isnt|not|super)(?!\\s*:)\\b)'
```
Comment (50Wliu) [inline grammars/coffeescript.cson]: ?: isn't needed

File: spec/coffee-script-spec.coffee [hunk 2]
```diff
@@ -178,6 +178,27 @@ describe "CoffeeScript grammar", ->
     expect(tokens[2]).toEqual value: "==", scopes: ["source.coffee", "keyword.operator.coffee"]
     expect(tokens[3]).toEqual value: " b", scopes: ["source.coffee"]
 
+  it "should tokenizes compound operators properly", ->
```
Comment (50Wliu) [inline spec/coffee-script-spec.coffee]: How about `it "tokenizes ..."`?

**Comment by 50Wliu** [comment]
Looks good at first glance to me, but I'll look at the regex in detail in a few hours.

**Comment by adrianlee44** [comment]
Updated this PR.

File: spec/coffee-script-spec.coffee [hunk 3]
```diff
@@ -178,6 +178,27 @@ describe "CoffeeScript grammar", ->
...
+      expect(tokens[0]).toEqual value: compoundOperator, scopes: ["source.coffee", "keyword.operator.coffee"]
+
+  it "tokenizes operators properly", ->
+    operators = ["!", "%", "^", "*", "/", "~", "?", ":", "-", "--", "+", "++", "<", ">", "&", "&&", "..", "...", "|", "||", "instanceof", "new", "delete", "typeof", "and", "or", "is", "isnt", "not", "super"]
+
+    for operator in operators
+      {tokens} = grammar.tokenizeLine(operator)
+      expect(tokens[0]).toEqual value: operator, scopes: ["source.coffee", "keyword.operator.coffee"]
+
+  it "does not tokenize as operators", ->
```
Comment (50Wliu) [inline spec/coffee-script-spec.coffee]: I think you forgot a word or two here :stuck_out_tongue:

File: spec/coffee-script-spec.coffee [hunk 4]
```diff
@@ -178,6 +178,27 @@ describe "CoffeeScript grammar", ->
...
+    notOperators = ["(/=", "-->", "=>"]
+
+    for notOperator in notOperators
+      {tokens} = grammar.tokenizeLine(notOperator)
+      expect(tokens[0]).toNotEqual value: notOperator, scopes: ["source.coffee", "keyword.operator.coffee"]
```
Comment (50Wliu) [inline spec/coffee-script-spec.coffee]: :art: can you change this to `.not.toEqual` please?

**Comment by 50Wliu** [comment]
Couldn't find anything wrong with the regex changes.  Just some minor spec nits, and then I think this is ready to :ship:!

File: grammars/coffeescript.cson [hunk 5]
```diff
@@ -143,7 +143,7 @@
     'name': 'keyword.control.coffee'
   }
   {
-    'match': 'and=|or=|!|%|&|\\^|\\*|\\/|(\\-)?\\-(?!>)|\\+\\+|\\+|~|==|=(?!>)|!=|<=|>=|<<=|>>=|>>>=|<>|<|>|!|&&|\\.\\.(\\.)?|\\?|\\||\\|\\||\\:|\\*=|(?<!\\()/=|%=|\\+=|\\-=|&=|\\^=|\\b(?<![\\.\\$])(instanceof|new|delete|typeof|and|or|is|isnt|not|super)(?!\\s*:)\\b'
+    'match': '((and|or|<<|>>>?|(?<!\\()\\/|[=!<>*%+\\-&^])?=(?!>)|[!%^*\\/~?:]|\\-?\\-(?!>)|\\+\\+?|<>|<|>|&&?|\\.\\.\\.?|\\|\\|?|\\b(?<![\\.\\$])(instanceof|new|delete|typeof|and|or|is|isnt|not|super)(?!\\s*:)\\b)'
```
Comment (50Wliu) [inline grammars/coffeescript.cson]: I don't think the outermost parentheses are needed.

File: spec/coffee-script-spec.coffee (hunk 4 above)
Comment (adrianlee44) [inline spec/coffee-script-spec.coffee]: :+1: yup

**Comment by adrianlee44** [comment]
Updated PR based on the comments. @50Wliu, thanks for checking the regex.

**Pull request merged**